// vafast/dist/utils/validators/validators.js
// src/utils/validators/validators.ts
import { Type } from "@sinclair/typebox";
import { TypeCompiler } from "@sinclair/typebox/compiler";
var compilerCache = /* @__PURE__ */ new WeakMap();
function getCompiledValidator(schema) {
  let compiler = compilerCache.get(schema);
  if (!compiler) {
    compiler = TypeCompiler.Compile(schema);
    compilerCache.set(schema, compiler);
  }
  return compiler;
}
function validateSchema(schema, data) {
  try {
    const compiler = getCompiledValidator(schema);
    if (compiler.Check(data)) {
      return { success: true, data };
    }
    const errors = [];
    for (const error of compiler.Errors(data)) {
      errors.push({
        path: error.path,
        message: error.message,
        code: "VALIDATION_FAILED",
        value: error.value
      });
    }
    return { success: false, errors };
  } catch (error) {
    return {
      success: false,
      errors: [
        {
          path: "",
          message: error instanceof Error ? error.message : "\u9A8C\u8BC1\u5F02\u5E38",
          code: "VALIDATION_EXCEPTION"
        }
      ]
    };
  }
}
function validateSchemaOrThrow(schema, data, context) {
  const compiler = getCompiledValidator(schema);
  if (!compiler.Check(data)) {
    throw new Error(`${context}\u9A8C\u8BC1\u5931\u8D25`);
  }
  return data;
}
function validateFast(schema, data) {
  const compiler = getCompiledValidator(schema);
  return compiler.Check(data);
}
function validateAllSchemas(config, data) {
  if (config.body) {
    validateSchemaOrThrow(config.body, data.body, "\u8BF7\u6C42\u4F53");
  }
  if (config.query) {
    validateSchemaOrThrow(config.query, data.query, "Query\u53C2\u6570");
  }
  if (config.params) {
    validateSchemaOrThrow(config.params, data.params, "\u8DEF\u5F84\u53C2\u6570");
  }
  if (config.headers) {
    validateSchemaOrThrow(config.headers, data.headers, "\u8BF7\u6C42\u5934");
  }
  if (config.cookies) {
    validateSchemaOrThrow(config.cookies, data.cookies, "Cookie");
  }
  return data;
}
function precompileSchemas(config) {
  if (config.body) getCompiledValidator(config.body);
  if (config.query) getCompiledValidator(config.query);
  if (config.params) getCompiledValidator(config.params);
  if (config.headers) getCompiledValidator(config.headers);
  if (config.cookies) getCompiledValidator(config.cookies);
}
function createValidator(schema) {
  return (data) => validateSchema(schema, data);
}
function getValidatorCacheStats() {
  return {
    cacheType: "WeakMap",
    note: "WeakMap \u7F13\u5B58\u4F1A\u968F Schema \u5BF9\u8C61\u81EA\u52A8\u6E05\u7406\uFF0C\u65E0\u5185\u5B58\u6CC4\u6F0F\u98CE\u9669"
  };
}
export {
  Type,
  createValidator,
  getValidatorCacheStats,
  precompileSchemas,
  validateAllSchemas,
  validateFast,
  validateSchema,
  validateSchemaOrThrow
};
