// vafast/dist/utils/formats.js
// src/utils/formats.ts
import { FormatRegistry } from "@sinclair/typebox";
var EMAIL_REGEX = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
var UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
var UUID_ANY_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
var CUID_REGEX = /^c[^\s-]{8,}$/i;
var CUID2_REGEX = /^[0-9a-z]+$/;
var ULID_REGEX = /^[0-9A-HJKMNP-TV-Z]{26}$/i;
var NANOID_REGEX = /^[A-Za-z0-9_-]{21}$/;
var URL_REGEX = /^https?:\/\/(?:www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b[-a-zA-Z0-9()@:%_+.~#?&/=]*$/;
var IPV4_REGEX = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
var IPV6_REGEX = /^(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$|^::(?:[0-9a-fA-F]{1,4}:){0,6}[0-9a-fA-F]{1,4}$|^(?:[0-9a-fA-F]{1,4}:){1,7}:$|^(?:[0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}$|^(?:[0-9a-fA-F]{1,4}:){1,5}(?::[0-9a-fA-F]{1,4}){1,2}$|^(?:[0-9a-fA-F]{1,4}:){1,4}(?::[0-9a-fA-F]{1,4}){1,3}$|^(?:[0-9a-fA-F]{1,4}:){1,3}(?::[0-9a-fA-F]{1,4}){1,4}$|^(?:[0-9a-fA-F]{1,4}:){1,2}(?::[0-9a-fA-F]{1,4}){1,5}$|^[0-9a-fA-F]{1,4}:(?::[0-9a-fA-F]{1,4}){1,6}$/;
var CIDR_REGEX = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/(?:3[0-2]|[12]?[0-9])$/;
var DATE_REGEX = /^\d{4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|[12][0-9]|3[01])$/;
var TIME_REGEX = /^(?:[01][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](?:\.\d{1,3})?$/;
var DATE_TIME_REGEX = /^\d{4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|[12][0-9]|3[01])T(?:[01][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](?:\.\d{1,3})?(?:Z|[+-](?:[01][0-9]|2[0-3]):[0-5][0-9])?$/;
var DURATION_REGEX = /^P(?:\d+Y)?(?:\d+M)?(?:\d+W)?(?:\d+D)?(?:T(?:\d+H)?(?:\d+M)?(?:\d+(?:\.\d+)?S)?)?$/;
var HOSTNAME_REGEX = /^(?=.{1,253}$)(?:(?!-)[a-zA-Z0-9-]{1,63}(?<!-)\.)*(?!-)[a-zA-Z0-9-]{1,63}(?<!-)$/;
var PHONE_CN_REGEX = /^1[3-9]\d{9}$/;
var PHONE_E164_REGEX = /^\+[1-9]\d{6,14}$/;
var OBJECTID_REGEX = /^[0-9a-fA-F]{24}$/;
var HEX_COLOR_REGEX = /^#(?:[0-9a-fA-F]{3}){1,2}$/;
var RGB_COLOR_REGEX = /^rgba?\(\s*(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\s*,\s*(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\s*,\s*(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(?:\s*,\s*(?:0|1|0?\.\d+))?\s*\)$/;
var BASE64_REGEX = /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/;
var BASE64URL_REGEX = /^[A-Za-z0-9_-]+$/;
var JWT_REGEX = /^[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+$/;
var EMOJI_REGEX = /^[\p{Emoji}]+$/u;
var SLUG_REGEX = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
var SEMVER_REGEX = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/;
function isValidCreditCard(value) {
  const digits = value.replace(/\D/g, "");
  if (digits.length < 13 || digits.length > 19) return false;
  let sum = 0;
  let isEven = false;
  for (let i = digits.length - 1; i >= 0; i--) {
    let digit = parseInt(digits[i], 10);
    if (isEven) {
      digit *= 2;
      if (digit > 9) digit -= 9;
    }
    sum += digit;
    isEven = !isEven;
  }
  return sum % 10 === 0;
}
var isRegistered = false;
function registerFormats() {
  if (isRegistered) return;
  isRegistered = true;
  FormatRegistry.Set("email", (v) => EMAIL_REGEX.test(v));
  FormatRegistry.Set("uuid", (v) => UUID_REGEX.test(v));
  FormatRegistry.Set("uuid-any", (v) => UUID_ANY_REGEX.test(v));
  FormatRegistry.Set("cuid", (v) => CUID_REGEX.test(v));
  FormatRegistry.Set("cuid2", (v) => CUID2_REGEX.test(v) && v.length >= 1);
  FormatRegistry.Set("ulid", (v) => ULID_REGEX.test(v));
  FormatRegistry.Set("nanoid", (v) => NANOID_REGEX.test(v));
  FormatRegistry.Set("objectid", (v) => OBJECTID_REGEX.test(v));
  FormatRegistry.Set("slug", (v) => SLUG_REGEX.test(v));
  FormatRegistry.Set("url", (v) => URL_REGEX.test(v));
  FormatRegistry.Set("uri", (v) => URL_REGEX.test(v));
  FormatRegistry.Set("ipv4", (v) => IPV4_REGEX.test(v));
  FormatRegistry.Set("ipv6", (v) => IPV6_REGEX.test(v));
  FormatRegistry.Set("ip", (v) => IPV4_REGEX.test(v) || IPV6_REGEX.test(v));
  FormatRegistry.Set("cidr", (v) => CIDR_REGEX.test(v));
  FormatRegistry.Set("hostname", (v) => HOSTNAME_REGEX.test(v));
  FormatRegistry.Set("date", (v) => DATE_REGEX.test(v));
  FormatRegistry.Set("time", (v) => TIME_REGEX.test(v));
  FormatRegistry.Set("date-time", (v) => DATE_TIME_REGEX.test(v));
  FormatRegistry.Set("datetime", (v) => DATE_TIME_REGEX.test(v));
  FormatRegistry.Set("duration", (v) => DURATION_REGEX.test(v));
  FormatRegistry.Set("phone", (v) => PHONE_CN_REGEX.test(v));
  FormatRegistry.Set("phone-cn", (v) => PHONE_CN_REGEX.test(v));
  FormatRegistry.Set("phone-e164", (v) => PHONE_E164_REGEX.test(v));
  FormatRegistry.Set("base64", (v) => BASE64_REGEX.test(v));
  FormatRegistry.Set("base64url", (v) => BASE64URL_REGEX.test(v));
  FormatRegistry.Set("jwt", (v) => JWT_REGEX.test(v));
  FormatRegistry.Set("hex-color", (v) => HEX_COLOR_REGEX.test(v));
  FormatRegistry.Set("rgb-color", (v) => RGB_COLOR_REGEX.test(v));
  FormatRegistry.Set(
    "color",
    (v) => HEX_COLOR_REGEX.test(v) || RGB_COLOR_REGEX.test(v)
  );
  FormatRegistry.Set("emoji", (v) => EMOJI_REGEX.test(v));
  FormatRegistry.Set("semver", (v) => SEMVER_REGEX.test(v));
  FormatRegistry.Set("credit-card", isValidCreditCard);
}
function registerFormat(name, validator) {
  FormatRegistry.Set(name, validator);
}
function hasFormat(name) {
  return FormatRegistry.Has(name);
}
var Patterns = {
  EMAIL: EMAIL_REGEX,
  UUID: UUID_REGEX,
  URL: URL_REGEX,
  IPV4: IPV4_REGEX,
  IPV6: IPV6_REGEX,
  DATE: DATE_REGEX,
  TIME: TIME_REGEX,
  DATE_TIME: DATE_TIME_REGEX,
  PHONE_CN: PHONE_CN_REGEX,
  PHONE_E164: PHONE_E164_REGEX,
  OBJECTID: OBJECTID_REGEX,
  HEX_COLOR: HEX_COLOR_REGEX,
  SLUG: SLUG_REGEX,
  SEMVER: SEMVER_REGEX,
  JWT: JWT_REGEX
};
export {
  Patterns,
  hasFormat,
  registerFormat,
  registerFormats
};
