// vafast/dist/utils/parsers.js
// src/utils/parsers.ts
import qs from "qs";
import cookie from "cookie";
async function parseBody(req) {
  const contentType = req.headers.get("content-type") || "";
  if (contentType.includes("application/json")) {
    return await req.json();
  }
  if (contentType.includes("application/x-www-form-urlencoded")) {
    const text = await req.text();
    return Object.fromEntries(new URLSearchParams(text));
  }
  return await req.text();
}
async function parseMultipartFormData(req) {
  const formData = await req.formData();
  const result = {
    fields: {},
    files: {}
  };
  for (const [key, value] of formData.entries()) {
    if (typeof value === "object" && value !== null && "name" in value && "type" in value && "size" in value) {
      const file = value;
      const arrayBuffer = await file.arrayBuffer();
      result.files[key] = {
        name: file.name,
        type: file.type,
        size: file.size,
        data: arrayBuffer
      };
    } else {
      result.fields[key] = value;
    }
  }
  return result;
}
async function parseBodyAs(req) {
  const body = await parseBody(req);
  return body;
}
async function parseFormData(req) {
  const contentType = req.headers.get("content-type") || "";
  if (!contentType.includes("multipart/form-data")) {
    throw new Error("\u8BF7\u6C42\u4E0D\u662F multipart/form-data \u683C\u5F0F");
  }
  return await parseMultipartFormData(req);
}
async function parseFile(req) {
  const contentType = req.headers.get("content-type") || "";
  if (!contentType.includes("multipart/form-data")) {
    throw new Error("\u8BF7\u6C42\u4E0D\u662F multipart/form-data \u683C\u5F0F");
  }
  const formData = await parseMultipartFormData(req);
  const fileKeys = Object.keys(formData.files);
  if (fileKeys.length === 0) {
    throw new Error("\u8BF7\u6C42\u4E2D\u6CA1\u6709\u6587\u4EF6");
  }
  if (fileKeys.length > 1) {
    throw new Error("\u8BF7\u6C42\u4E2D\u5305\u542B\u591A\u4E2A\u6587\u4EF6\uFF0C\u8BF7\u4F7F\u7528 parseFormData");
  }
  return formData.files[fileKeys[0]];
}
function extractQueryString(url) {
  const qIndex = url.indexOf("?");
  if (qIndex === -1) return "";
  const hashIndex = url.indexOf("#", qIndex);
  return hashIndex === -1 ? url.substring(qIndex + 1) : url.substring(qIndex + 1, hashIndex);
}
function parseQuery(req) {
  const queryString = extractQueryString(req.url);
  if (!queryString) return {};
  return qs.parse(queryString);
}
function parseQueryFast(req) {
  const queryString = extractQueryString(req.url);
  if (!queryString) return {};
  const result = /* @__PURE__ */ Object.create(null);
  const pairs = queryString.split("&");
  for (const pair of pairs) {
    const eqIndex = pair.indexOf("=");
    if (eqIndex === -1) {
      result[decodeURIComponent(pair)] = "";
    } else {
      const key = decodeURIComponent(pair.substring(0, eqIndex));
      const value = decodeURIComponent(pair.substring(eqIndex + 1));
      result[key] = value;
    }
  }
  return result;
}
function parseHeaders(req) {
  const headers = /* @__PURE__ */ Object.create(null);
  req.headers.forEach((value, key) => {
    headers[key] = value;
  });
  return headers;
}
function getHeader(req, name) {
  return req.headers.get(name);
}
function parseCookies(req) {
  const cookieHeader = req.headers.get("cookie");
  if (!cookieHeader) return {};
  try {
    const parsed = cookie.parse(cookieHeader);
    const result = {};
    for (const [key, value] of Object.entries(parsed)) {
      if (value !== void 0 && value !== null) {
        result[key] = value;
      }
    }
    return result;
  } catch {
    return {};
  }
}
function parseCookiesFast(req) {
  const cookieHeader = req.headers.get("cookie");
  if (!cookieHeader) return {};
  const result = /* @__PURE__ */ Object.create(null);
  const pairs = cookieHeader.split(";");
  for (const pair of pairs) {
    const trimmed = pair.trim();
    const eqIndex = trimmed.indexOf("=");
    if (eqIndex > 0) {
      const key = trimmed.substring(0, eqIndex).trim();
      const value = trimmed.substring(eqIndex + 1).trim();
      result[key] = value.startsWith('"') && value.endsWith('"') ? value.slice(1, -1) : value;
    }
  }
  return result;
}
function getCookie(req, name) {
  const cookieHeader = req.headers.get("cookie");
  if (!cookieHeader) return null;
  const prefix = `${name}=`;
  const pairs = cookieHeader.split(";");
  for (const pair of pairs) {
    const trimmed = pair.trim();
    if (trimmed.startsWith(prefix)) {
      const value = trimmed.substring(prefix.length).trim();
      return value.startsWith('"') && value.endsWith('"') ? value.slice(1, -1) : value;
    }
  }
  return null;
}
export {
  getCookie,
  getHeader,
  parseBody,
  parseBodyAs,
  parseCookies,
  parseCookiesFast,
  parseFile,
  parseFormData,
  parseHeaders,
  parseQuery,
  parseQueryFast
};
