// vafast/dist/utils/response.js
// src/utils/response.ts
function json(data, status = 200, headers = {}) {
  const body = JSON.stringify(data);
  if (Object.keys(headers).length === 0) {
    return new Response(body, {
      status,
      headers: { "Content-Type": "application/json" }
    });
  }
  const h = new Headers({
    "Content-Type": "application/json",
    ...headers
  });
  return new Response(body, {
    status,
    headers: h
  });
}
var JSON_HEADERS = { "Content-Type": "application/json" };
var TEXT_HEADERS = { "Content-Type": "text/plain" };
function mapResponse(response) {
  if (response instanceof Response) return response;
  switch (response?.constructor?.name) {
    case "String":
      return new Response(response, { headers: TEXT_HEADERS });
    case "Object":
    case "Array":
      return new Response(JSON.stringify(response), { headers: JSON_HEADERS });
    case "Number":
    case "Boolean":
      return new Response(String(response), { headers: TEXT_HEADERS });
    case void 0:
      return new Response(null, { status: 204 });
    case "ReadableStream":
      return new Response(response);
    case "Blob":
      return new Response(response);
    case "ArrayBuffer":
      return new Response(response);
    case "Uint8Array":
      return new Response(response);
    default:
      if (response instanceof Promise) {
        return response.then(mapResponse);
      }
      return new Response(JSON.stringify(response), { headers: JSON_HEADERS });
  }
}
function redirect(location, status = 302) {
  return new Response(null, {
    status,
    headers: {
      Location: location
    }
  });
}
function text(content, status = 200, headers = {}) {
  const h = new Headers({
    "Content-Type": "text/plain; charset=utf-8",
    ...headers
  });
  return new Response(content, {
    status,
    headers: h
  });
}
function html(content, status = 200, headers = {}) {
  const h = new Headers({
    "Content-Type": "text/html; charset=utf-8",
    ...headers
  });
  return new Response(content, {
    status,
    headers: h
  });
}
function empty(status = 204, headers = {}) {
  return new Response(null, {
    status,
    headers
  });
}
function stream(stream2, status = 200, headers = {}) {
  const h = new Headers({
    "Content-Type": "application/octet-stream",
    ...headers
  });
  return new Response(stream2, {
    status,
    headers: h
  });
}
export {
  empty,
  html,
  json,
  mapResponse,
  redirect,
  stream,
  text
};
