// vafast/dist/utils/route-registry.js
// src/utils/route-registry.ts
var RouteRegistry = class {
  /** 所有路由元信息 */
  routes = [];
  /** 路由映射表：METHOD:fullPath -> RouteMeta */
  routeMap = /* @__PURE__ */ new Map();
  /** 分类映射表：category -> RouteMeta[] */
  categoryMap = /* @__PURE__ */ new Map();
  constructor(routes) {
    this.buildRegistry(routes);
  }
  /**
   * 构建注册表
   */
  buildRegistry(routes) {
    for (const route of routes) {
      const meta = {
        method: route.method,
        path: route.path,
        fullPath: route.fullPath,
        name: route.name,
        description: route.description
      };
      for (const key of Object.keys(route)) {
        if (!["method", "path", "fullPath", "name", "description", "handler", "middleware", "middlewareChain"].includes(key)) {
          meta[key] = route[key];
        }
      }
      this.routes.push(meta);
      this.routeMap.set(`${route.method}:${route.fullPath}`, meta);
      const category = this.extractCategory(route.fullPath);
      if (!this.categoryMap.has(category)) {
        this.categoryMap.set(category, []);
      }
      this.categoryMap.get(category).push(meta);
    }
  }
  /**
   * 提取分类（第一段路径）
   */
  extractCategory(path) {
    const segments = path.split("/").filter(Boolean);
    return segments[0] || "root";
  }
  // ============================================
  // 查询接口
  // ============================================
  /**
   * 获取所有路由元信息
   */
  getAll() {
    return [...this.routes];
  }
  /**
   * 按 method + path 查询路由
   */
  get(method, path) {
    return this.routeMap.get(`${method}:${path}`);
  }
  /**
   * 检查路由是否存在
   */
  has(method, path) {
    return this.routeMap.has(`${method}:${path}`);
  }
  /**
   * 按分类获取路由
   */
  getByCategory(category) {
    return this.categoryMap.get(category) || [];
  }
  /**
   * 获取所有分类
   */
  getCategories() {
    return Array.from(this.categoryMap.keys()).sort();
  }
  /**
   * 筛选有特定字段的路由
   *
   * @example
   * ```typescript
   * // 获取所有配置了 webhook 的路由
   * const webhookRoutes = registry.filter('webhook')
   * ```
   */
  filter(field) {
    return this.routes.filter((r) => field in r && r[field] !== void 0);
  }
  /**
   * 按条件筛选路由
   *
   * @example
   * ```typescript
   * // 获取所有 POST 请求
   * const postRoutes = registry.filterBy(r => r.method === 'POST')
   * ```
   */
  filterBy(predicate) {
    return this.routes.filter(predicate);
  }
  /**
   * 获取路由数量
   */
  get size() {
    return this.routes.length;
  }
  /**
   * 遍历所有路由
   */
  forEach(callback) {
    this.routes.forEach(callback);
  }
  /**
   * 映射所有路由
   */
  map(callback) {
    return this.routes.map(callback);
  }
};
function createRouteRegistry(routes) {
  return new RouteRegistry(routes);
}
var globalRegistry = null;
function setGlobalRegistry(registry) {
  globalRegistry = registry;
}
function getRouteRegistry() {
  if (!globalRegistry) {
    throw new Error("RouteRegistry not initialized. Make sure Server is created first.");
  }
  return globalRegistry;
}
function getRoute(method, path) {
  return getRouteRegistry().get(method, path);
}
function getAllRoutes() {
  return getRouteRegistry().getAll();
}
function filterRoutes(field) {
  return getRouteRegistry().filter(field);
}
export {
  RouteRegistry,
  createRouteRegistry,
  filterRoutes,
  getAllRoutes,
  getRoute,
  getRouteRegistry,
  setGlobalRegistry
};
