// vafast/dist/router/radix-tree.js
// src/router/radix-tree.ts
var RadixRouter = class {
  root;
  constructor() {
    this.root = this.createNode("");
  }
  createNode(path) {
    return {
      path,
      children: /* @__PURE__ */ Object.create(null),
      handlers: /* @__PURE__ */ Object.create(null)
    };
  }
  /** 分割路径 */
  splitPath(path) {
    return path.split("/").filter(Boolean);
  }
  /** 编译器函数 - 用于预编译中间件链 */
  compiler;
  /** 设置中间件编译器 */
  setCompiler(compiler) {
    this.compiler = compiler;
  }
  /** 注册路由 */
  register(method, pattern, handler, middleware = []) {
    const segments = this.splitPath(pattern);
    let node = this.root;
    for (const segment of segments) {
      const firstChar = segment[0];
      if (firstChar === ":") {
        if (!node.paramChild) {
          node.paramChild = this.createNode(segment);
          node.paramChild.paramName = segment.substring(1);
        }
        node = node.paramChild;
      } else if (firstChar === "*") {
        if (!node.wildcardChild) {
          node.wildcardChild = this.createNode(segment);
          node.wildcardChild.paramName = segment.length > 1 ? segment.substring(1) : "*";
        }
        node = node.wildcardChild;
        break;
      } else {
        if (!node.children[segment]) {
          node.children[segment] = this.createNode(segment);
        }
        node = node.children[segment];
      }
    }
    const routeHandler = { handler, middleware };
    if (this.compiler && middleware.length === 0) {
      routeHandler.compiled = this.compiler([], handler);
    }
    node.handlers[method] = routeHandler;
  }
  /** 预编译所有路由（在添加全局中间件后调用） */
  precompileAll(globalMiddleware) {
    if (!this.compiler) return;
    this.precompileNode(this.root, globalMiddleware);
  }
  precompileNode(node, globalMiddleware) {
    for (const method in node.handlers) {
      const routeHandler = node.handlers[method];
      if (routeHandler) {
        const allMiddleware = [...globalMiddleware, ...routeHandler.middleware];
        routeHandler.compiled = this.compiler(
          allMiddleware,
          routeHandler.handler
        );
      }
    }
    for (const key in node.children) {
      this.precompileNode(node.children[key], globalMiddleware);
    }
    if (node.paramChild) {
      this.precompileNode(node.paramChild, globalMiddleware);
    }
    if (node.wildcardChild) {
      this.precompileNode(node.wildcardChild, globalMiddleware);
    }
  }
  /** 匹配路由 */
  match(method, path) {
    const segments = this.splitPath(path);
    const params = /* @__PURE__ */ Object.create(null);
    const node = this.matchNode(this.root, segments, 0, params);
    if (!node) return null;
    const routeHandler = node.handlers[method];
    if (!routeHandler) return null;
    return {
      handler: routeHandler.handler,
      middleware: routeHandler.middleware,
      params,
      compiled: routeHandler.compiled
    };
  }
  /** 递归匹配节点 (优先级: 静态 > 动态参数 > 通配符) */
  matchNode(node, segments, index, params) {
    if (index === segments.length) {
      for (const method in node.handlers) {
        if (node.handlers[method]) return node;
      }
      return null;
    }
    const segment = segments[index];
    const staticChild = node.children[segment];
    if (staticChild) {
      const result = this.matchNode(staticChild, segments, index + 1, params);
      if (result) return result;
    }
    if (node.paramChild) {
      const paramName = node.paramChild.paramName;
      const oldValue = params[paramName];
      params[paramName] = segment;
      const result = this.matchNode(
        node.paramChild,
        segments,
        index + 1,
        params
      );
      if (result) return result;
      if (oldValue === void 0) {
        delete params[paramName];
      } else {
        params[paramName] = oldValue;
      }
    }
    if (node.wildcardChild) {
      params[node.wildcardChild.paramName || "*"] = segments.slice(index).join("/");
      return node.wildcardChild;
    }
    return null;
  }
  /** 获取路径允许的 HTTP 方法 */
  getAllowedMethods(path) {
    const segments = this.splitPath(path);
    const node = this.findNode(segments);
    if (!node) return [];
    const methods = [];
    for (const method in node.handlers) {
      if (node.handlers[method]) {
        methods.push(method);
      }
    }
    return methods;
  }
  /** 查找节点（不提取参数） */
  findNode(segments) {
    let node = this.root;
    for (const segment of segments) {
      if (node.children[segment]) {
        node = node.children[segment];
      } else if (node.paramChild) {
        node = node.paramChild;
      } else if (node.wildcardChild) {
        return node.wildcardChild;
      } else {
        return null;
      }
    }
    return node;
  }
  /** 获取所有已注册的路由 */
  getRoutes() {
    const routes = [];
    this.collectRoutes(this.root, "", routes);
    return routes;
  }
  collectRoutes(node, prefix, routes) {
    const currentPath = prefix + (node.path ? "/" + node.path : "");
    for (const method in node.handlers) {
      if (node.handlers[method]) {
        routes.push({ method, path: currentPath || "/" });
      }
    }
    for (const key in node.children) {
      this.collectRoutes(node.children[key], currentPath, routes);
    }
    if (node.paramChild) {
      this.collectRoutes(node.paramChild, currentPath, routes);
    }
    if (node.wildcardChild) {
      this.collectRoutes(node.wildcardChild, currentPath, routes);
    }
  }
};
export {
  RadixRouter
};
