// vafast/dist/middleware/authMiddleware.js
// src/middleware.ts
var VafastError = class extends Error {
  status;
  type;
  expose;
  constructor(message, options = {}) {
    super(message);
    this.name = "VafastError";
    this.status = options.status ?? 500;
    this.type = options.type ?? "internal_error";
    this.expose = options.expose ?? false;
    if (options.cause) this.cause = options.cause;
  }
};

// src/utils/parsers.ts
import qs from "qs";
import cookie from "cookie";
function parseCookies(req) {
  const cookieHeader = req.headers.get("cookie");
  if (!cookieHeader) return {};
  try {
    const parsed = cookie.parse(cookieHeader);
    const result = {};
    for (const [key, value] of Object.entries(parsed)) {
      if (value !== void 0 && value !== null) {
        result[key] = value;
      }
    }
    return result;
  } catch {
    return {};
  }
}

// src/utils/handle.ts
function getCookie(req, key) {
  const cookies = parseCookies(req);
  return cookies[key] || null;
}

// src/middleware/authMiddleware.ts
var requireAuth = async (req, next) => {
  const token = getCookie(req, "auth");
  if (!token || token !== "valid-token") {
    throw new VafastError("Unauthorized", {
      status: 401,
      type: "unauthorized",
      expose: true
    });
  }
  return next();
};
export {
  requireAuth
};
