// vafast/dist/middleware/component-renderer.js
// src/middleware/component-renderer.ts
var renderVueSSR = async (componentImport, req, preloadedDeps) => {
  try {
    const { createSSRApp, renderToString } = preloadedDeps || await Promise.all([import("vue"), import("@vue/server-renderer")]).then(
      ([vue, renderer]) => ({
        createSSRApp: vue.createSSRApp,
        renderToString: renderer.renderToString
      })
    );
    const componentModule = await componentImport();
    const component = componentModule.default || componentModule;
    const app = createSSRApp(component);
    app.provide("routeInfo", {
      params: req.params || {},
      query: Object.fromEntries(new URL(req.url).searchParams),
      pathname: new URL(req.url).pathname
    });
    const html = await renderToString(app);
    return new Response(
      `
      <!doctype html>
      <html>
        <head>
          <meta charset="utf-8">
          <title>Vue SSR App</title>
        </head>
        <body>
          <div id="app">${html}</div>
          <script>
            window.__ROUTE_INFO__ = {
              params: ${JSON.stringify(req.params || {})},
              query: ${JSON.stringify(Object.fromEntries(new URL(req.url).searchParams))},
              pathname: '${new URL(req.url).pathname}'
            };
          </script>
          <script type="module" src="/client.js"></script>
        </body>
      </html>
    `,
      {
        headers: { "Content-Type": "text/html; charset=utf-8" }
      }
    );
  } catch (error) {
    console.error("Vue SSR \u6E32\u67D3\u5931\u8D25:", error);
    return new Response(
      `
      <!doctype html>
      <html>
        <head><title>\u6E32\u67D3\u9519\u8BEF</title></head>
        <body>
          <h1>Vue SSR \u6E32\u67D3\u5931\u8D25</h1>
          <p>\u9519\u8BEF\u4FE1\u606F: ${error instanceof Error ? error.message : "\u672A\u77E5\u9519\u8BEF"}</p>
        </body>
      </html>
    `,
      { status: 500 }
    );
  }
};
var renderReactSSR = async (componentImport, req, preloadedDeps) => {
  try {
    const { createElement, renderToString } = preloadedDeps || await Promise.all([import("react"), import("react-dom/server")]).then(
      ([react, renderer]) => ({
        createElement: react.createElement,
        renderToString: renderer.renderToString
      })
    );
    const componentModule = await componentImport();
    const Component = componentModule.default || componentModule;
    const content = createElement(Component, {
      req,
      params: req.params || {},
      query: Object.fromEntries(new URL(req.url).searchParams)
    });
    const html = renderToString(content);
    return new Response(
      `
      <!doctype html>
      <html>
        <head>
          <meta charset="utf-8">
          <title>React SSR App</title>
        </head>
        <body>
          <div id="root">${html}</div>
          <script>
            window.__ROUTE_INFO__ = {
              params: ${JSON.stringify(req.params || {})},
              query: ${JSON.stringify(Object.fromEntries(new URL(req.url).searchParams))},
              pathname: '${new URL(req.url).pathname}'
            };
          </script>
          <script type="module" src="/client.js"></script>
        </body>
      </html>
    `,
      {
        headers: { "Content-Type": "text/html; charset=utf-8" }
      }
    );
  } catch (error) {
    console.error("React SSR \u6E32\u67D3\u5931\u8D25:", error);
    return new Response(
      `
      <!doctype html>
      <html>
        <head><title>\u6E32\u67D3\u9519\u8BEF</title></head>
        <body>
          <h1>React SSR \u6E32\u67D3\u5931\u8D25</h1>
          <p>\u9519\u8BEF\u4FE1\u606F: ${error instanceof Error ? error.message : "\u672A\u77E5\u9519\u8BEF"}</p>
        </body>
      </html>
    `,
      { status: 500 }
    );
  }
};
var vueRenderer = (preloadedDeps) => {
  return async (req, next) => {
    req.renderVue = async (componentImport) => {
      return await renderVueSSR(componentImport, req, preloadedDeps);
    };
    return next();
  };
};
var reactRenderer = (preloadedDeps) => {
  return async (req, next) => {
    req.renderReact = async (componentImport) => {
      return await renderReactSSR(componentImport, req, preloadedDeps);
    };
    return next();
  };
};
export {
  reactRenderer,
  vueRenderer
};
