// vafast/dist/middleware/rateLimit.js
// src/middleware.ts
var VafastError = class extends Error {
  status;
  type;
  expose;
  constructor(message, options = {}) {
    super(message);
    this.name = "VafastError";
    this.status = options.status ?? 500;
    this.type = options.type ?? "internal_error";
    this.expose = options.expose ?? false;
    if (options.cause) this.cause = options.cause;
  }
};

// src/middleware/rateLimit.ts
var store = /* @__PURE__ */ new Map();
function rateLimit(options = {}) {
  const windowMs = options.windowMs ?? 6e4;
  const max = options.max ?? 30;
  const keyFn = options.keyFn ?? getIP;
  return async (req, next) => {
    const key = keyFn(req);
    const now = Date.now();
    const entry = store.get(key);
    if (entry && entry.expires > now) {
      if (entry.count >= max) {
        throw new VafastError("Too many requests", {
          status: 429,
          type: "rate_limit",
          expose: true
        });
      }
      entry.count += 1;
    } else {
      store.set(key, { count: 1, expires: now + windowMs });
    }
    return next();
  };
}
function getIP(req) {
  return req.headers.get("cf-connecting-ip") || // Cloudflare
  req.headers.get("x-forwarded-for")?.split(",")[0]?.trim() || // Vercel
  "unknown";
}
export {
  rateLimit
};
